### ISSUE TITLE
Module states cannot be fully synchronized due to the DDP broadcast_buffers breaking change

### ISSUE THREAD

**Issue by NcJie** [pytorch/pytorch#111138]
### 🐛 Describe the bug

From the recent breaking change in the `2.1.0` in the https://github.com/pytorch/pytorch/pull/100729, when the `broadcast_buffers` is set to `False`, it caused the DDP to not fully synchronize module states (i.e., only the parameter is synchronized but the buffer is being left out out-of-sync) during the DDP initialization and final step joining. This caused some issues in the following scenarios. 
1. With the random initialization, the master process module's parameter is broadcasted but not the buffer. While we can fix this by ensuring that all ranks are initialized with the same random seed, there are however some scenarios where we still want to use a different random seed for each rank. 
2. With the checkpoint loading, previously we only needed to load the checkpoint in the master process and states (parameters and buffers) will be broadcasted to all other processes. While we can fix this by loading the exact same checkpoint in each process, there's also a scenario in which not all processes would have direct access to the checkpoint, for example, in a multi-node configuration where only the master node has access to the local file system checkpoint. 

Also in the DDP [documentation](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel), it specifically states that the `broadcast_buffers` flag is used to control the buffer broadcasting at the beginning of a module forward function, but not during the initialization and module final joining. 
```
broadcast_buffers ([bool] – Flag that enables syncing (broadcasting) buffers of the module at the beginning of the forward function. (default: True)
```

The documentation states a different behavior from the new changes, and the new change also caused some issues mentioned above in the module loading in the DDP, is this an intended behavior or it's a bug? 

### Versions

2.1.0

**Comment by yang-song** [comment]
Agreed the documentation is confusing. Should explicitly state that `broadcast_buffers` controls buffer broadcasting at the initialization or module final joining, not "at the beginning of a module forward function", which can be interpreted as happening at each training step.

**Comment by wconstab** [comment]
I tend to agree that the PR https://github.com/pytorch/pytorch/pull/100729 broke the documented/established behavior and i'm not sure what the considerations were at the time for landing it.  We need to discuss what to do about it with @xing-liu, @mrshenli @fegin @rohan-varma
